Mutation

Tu-S opened the pull request.

Tu-S committed:
tried to define genome for classes

Tu-S committed:
genome upd

Tu-S committed:
mutate funcs

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/genome/GenomeUtils.java:
```diff
@@ -1,4 +1,59 @@
+     * Изменяем случайным образом параметры полосы
+     * @param laneGenome - полоса, которую изменяем
+     */
+    private static void mutateLane(LaneGenome laneGenome){
```
Давай ограничим сверху и снизу возможные значения скоростей. Надо помнить, что мутации происходят много раз и тут может получиться очень большое/маленькое значение.

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/genome/GenomeUtils.java:
```diff
@@ -1,4 +1,59 @@
+         var incr = maxSpeed + maxSpeed/MAX_SPEED_PART;
+         var decr = maxSpeed - maxSpeed/MIN_SPEED_PART;
+         var newSpeed = decr + Math.random()*(incr-decr);
+         laneGenome.setMaxSpeed(newSpeed);
```
Если меняем ограничения скорости по полосам, то надо будет придумать новую функцию. оценки. Сейчас она полагается на одинаковые ограничения, кроме того, там идет расчет avgSpeed/maxAllowedSpeed, то есть при том же движении если ограничение ниже, то оценка будет выше. Это скорее заметка, чтобы не забыть и придумать оценку.

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/genome/NodeGenome.java:
```diff
@@ -0,0 +1,52 @@
+    private TrafficLight copyTrafficLight(TrafficLight trafficLight) {
+        var res = new TrafficLight();
+        for (var c : trafficLight.getConfigs()) {
+            res.addConfig(new TrafficLightConfig(c.getDelay()));
```
Такие светофоры работать не будут, у них не указаны дороги в конфигах

SpykeX3 reviewed (changes requested):
В целом нормально, только пока не понятно, как это с нашими дорогами связано, нам же нужно используя какой-то слепок RoadMap и генома получить новый RoadMap. Возникает вопрос, где и в каком виде будут хранится геномы дорог и нод? Судя по тому, что у них только примитивные характеристики, они должны быть прилинкованы к соответсвующим дорогам и нодам снаружи. Как вариант можно положить эти геномы внутрь нод и дорог, изначально ставить значения из json-конфига и поменять геттеры, чтобы указывали на геном, а не на свои поля, потом сделать конструктор копирования для RoadMap, чтобы можно было параллельно использовать несколько карт.

Tu-S committed:
not important

Tu-S committed:
fixes 

Добавил функционал создания N экземпляров карты из конфига. Поменял классы генома.  Изменил функции мутации.

Tu-S committed:
copier upd

теперь можем копировать что угодно

Tu-S committed:
crossbreeding

Tu-S committed:
убирал машины из потомков

SpykeX3 committed:
change heatmap builder to work with lanes

Tu-S committed:
wrapper for map mutation

Commit:
Add selection

Commit:
Add selection test

Tu-S committed:
default cars

Tu-S committed:
set new periods in trafficlights

Tu-S committed:
algo


algo v1

Tu-S committed:
upd

Tu-S committed:
upd

Tu-S committed:
cur version

Tu-S committed:
upd

Commit:
No enter record fix

Tu-S committed:
clear map from cars

Commit:
Selection update

Tu-S committed:
upd + breeding bug somewhere

Tu-S committed:
upd

Tu-S committed:
o

Tu-S committed:
new crossbreeding

Tu-S committed:
bug fix

Commit:
Selection update

Commit:
Change surv. rate field type

Tu-S committed:
added: ids in maps, some differences in algo v2

Tu-S committed:
upd

Commit:
Refactor code

Commit:
Selection update

Commit:
Selection test update

Tu-S committed:
small upd
